README.md
# starparse [<img src="https://img.shields.io/gitlab/pipeline/alen/starparse/main?gitlab_url=https%3A%2F%2Fgitlab.home.alen.sh%2F&label=Gitlab%20CI&style=flat-square" align="right">](https://gitlab.home.alen.sh/alen/starparse) [<img src="https://img.shields.io/travis/buhanec/starparse/main.svg?label=Travis+CI&style=flat-square" align="right">](https://travis-ci.org/buhanec/starparse) [<img src="https://img.shields.io/azure-devops/build/buhanec/b18841f7-3636-4121-b15f-f924dc02c956/4/main?label=Azure%20DevOps&style=flat-square" align="right">](https://dev.azure.com/buhanec/starparse/_build)

Small Python project to parse save game files for Starbound.

Works with 1.0.x / nightly save files, both to unpack and re-pack.

Look at [`tests/test_e2e.py`](tests/test_e2e.py) for an example of reading a save file.


CHANGELOG.md
# Changelog

## v0.1.1 (2021-03-11)
- Make compatible with Python 3.6 and Python 3.10

## v0.1.0 (2021-03-10)
- Initially cleaned up and distributed


tests/__init__.py
from unittest import TestCase
from unittest.mock import patch
from collections import OrderedDict
from starparse import pack, unpack
from math import isnan
import random
import string


def parity(packer, unpacker, reference, asserter, packed_reference=None):
    packed = packer(reference)
    if packed_reference is not None:
        asserter(packed, packed_reference)
    unpacked, _ = unpacker(packed)
    asserter(unpacked, reference)


class StructTest(TestCase):
    def test(self):
        cases = [
            (('4c', b'Alen',),              ('Alen',    4)),
            (('4c', b'Alen B',),            ('Alen',    4)),
            (('2c', b'Alen B', 2),          ('en',      4)),
            (('b',  b'\x01\xFF',),          (1,         1)),
            (('b',  b'\x00\xFF',),          (0,         1)),
            (('<i', b'\x00\x88\x00\x00'),   (34816,     4))
        ]
        for args, expected in cases:
            self.assertEqual(unpack.struct(*args), expected)


class UintTest(TestCase):

    def test(self):
        cases = [
            ((b'\xFF\x00',),        (16256, 2)),
            ((b'\x01\xFF',),        (1,     1)),
            ((b'\x00\xFF',),        (0,     1)),
            ((b'A',),               (65,    1)),
            ((b'Alen',),            (65,    1)),
            ((b'\x88\x00',),        (1024,  2)),
            ((b'\x88\x00\x00',),    (1024,  2)),
            ((b'\x01',),            (1,     1)),
            ((b'Alen', 1),          (108,   2)),
            ((b'\x00\x88\x00', 1),  (1024,  3)),
            ((b'\x00\x00\x00', 1),  (0,     2))
        ]
        for args, expected in cases:
            self.assertEqual(unpack.uint(*args), expected)

    def test_parity(self):
        for n in range(0, 20000):
            parity(pack.uint, unpack.uint, n, self.assertEqual)


class IntTest(TestCase):

    def test(self):
        cases = [
            ((b'\xbf\xffA',),       (-524257,   3)),
            ((b'\x01\xFF',),        (-1,        1)),
            ((b'\x00\xFF',),        (0,         1)),
            ((b'A',),               (-33,       1)),
            ((b'Alen',),            (-33,       1)),
            ((b'\x88\x00',),        (512,       2)),
            ((b'\x88\x00\x00',),    (512,       2)),
            ((b'\x01',),            (-1,        1)),
            ((b'\xFF\x00',),        (8128,      2)),
            ((b'Alen', 1),          (54,        2)),
            ((b'\x00\x88\x00', 1),  (512,       3)),
            ((b'\x00\x00\x00', 1),  (0,         2))
        ]
        for args, expected in cases:
            self.assertEqual(unpack.int_(*args), expected)

    def test_parity(self):
        for n in range(-20000, 20000):
            parity(pack.int_, unpack.int_, n, self.assertEqual)


class StrTest(TestCase):

    def test_parity(self):
        characters = string.ascii_letters + string.digits + ' '
        for n in range(100):
            length = random.randint(0, 10)
            case = ''.join(random.choice(characters) for _ in range(length))
            parity(pack.str_, unpack.str_, case, self.assertEqual)


class BoolTest(TestCase):

    def test_parity(self):
        parity(pack.bool_, unpack.bool_, True, self.assertEqual)
        parity(pack.bool_, unpack.bool_, False, self.assertEqual)


class NoneTest(TestCase):

    def test_parity(self):
        parity(pack.none, unpack.none, None, self.assertEqual)


class FloatTest(TestCase):

    def test_parity(self):
        def assert_nans(a, b):
            self.assertTrue(isnan(a))
            self.assertTrue(isnan(b))

        def gen(start, stop, steps):
            step = (stop - start) / steps
            return [stop + i * step for i in range(steps)]

        cases = []
        cases.extend(gen(-1e20, 1e20, 2500))
        cases.extend(gen(-1, 1, 2500))
        cases.extend([float('inf'), -float('inf'), 0.0, -0.0])
        for case in cases:
            parity(pack.float_, unpack.float_, case, self.assertEqual)
        parity(pack.float_, unpack.float_, float('nan'), assert_nans)
        parity(pack.float_, unpack.float_, -float('nan'), assert_nans)


class TypeTest(TestCase):

    def test_ordered(self):
        pass

    @patch('starparse.config.ORDERED_DICT', False)
    def test_unordered(self):
        cases = [
            ((b'\x01',),    (None, 1)),
            ((b'\x02',),    (float, 1)),
            ((b'\x03',),    (bool, 1)),
            ((b'\x04',),    (int, 1)),
            ((b'\x05',),    (str, 1)),
            ((b'\x06',),    (list, 1)),
            ((b'\x07',),    (dict, 1)),
        ]
        for args, expected in cases:
            self.assertEqual(unpack.type_(*args), expected)

    @patch('starparse.config.ORDERED_DICT', True)
    def test_unordered(self):
        cases = [
            ((b'\x01',), (None, 1)),
            ((b'\x02',), (float, 1)),
            ((b'\x03',), (bool, 1)),
            ((b'\x04',), (int, 1)),
            ((b'\x05',), (str, 1)),
            ((b'\x06',), (list, 1)),
            ((b'\x07',), (dict, 1)),
        ]
        for args, expected in cases:
            self.assertEqual(unpack.type_(*args), expected)


class ListTest(TestCase):

    # TODO: Mock out other functions
    def test_parity(self):
        cases = [[], [1, 2, 3], ["a", "b", "c"], [1, "b", []]]
        for case in cases:
            parity(pack.list_, unpack.list_, case, self.assertEqual)


class DictTest(TestCase):

    # TODO: Mock out other functions
    @patch('starparse.config.ORDERED_DICT', False)
    def test_parity(self):
        cases = [{}, {"a": 1, "b": 2}, {"a": "a", "b": 2}, {"a": {"b": {}}}]
        for case in cases:
            parity(pack.dict_, unpack.dict_, case, self.assertEqual)

    # TODO: Mock out other functions
    @patch('starparse.config.ORDERED_DICT', True)
    def test_parity(self):
        cases = [
            OrderedDict({}),
            OrderedDict({"a": 1, "b": 2}),
            OrderedDict({"a": "a", "b": 2}),
            OrderedDict({"a": OrderedDict({"b": OrderedDict({})})})
        ]
        for case in cases:
            parity(pack.dict_, unpack.dict_, case, self.assertEqual)


tests/test_e2e.py
try:
    from importlib.resources import read_binary
except ModuleNotFoundError:
    from importlib_resources import read_binary
import itertools

import pytest

from starparse import config, pack, unpack
from tests import ci_players


@pytest.mark.parametrize('player, ordered_dict',
                         itertools.product(ci_players.PLAYERS, (True, False)))
def test_unordered(player, ordered_dict, monkeypatch):
    monkeypatch.setattr(config, 'ORDERED_DICT', ordered_dict)
    file = read_binary(ci_players, f'{player}.player')

    # Unpack
    save_format, entity, flags, offset = unpack.header(file, 0)
    unpacked, limit = unpack.typed(file, offset)

    assert limit == len(file)
    assert save_format == b'SBVJ01'

    # Pack
    packed = pack.header(save_format, entity, flags) + pack.typed(unpacked)

    assert len(packed) == len(file)

    # Re-unpack
    save_format_2, entity_2, flags_2, offset_2 = unpack.header(packed, 0)
    unpacked_2, limit_2 = unpack.typed(packed, offset_2)
    assert save_format == save_format_2
    assert entity == entity_2
    assert flags == flags_2
    assert offset == offset_2
    assert limit == limit_2
    assert unpacked == unpacked_2

    if ordered_dict:
        assert packed == file


tests/test_pack_unpack.py
from collections import OrderedDict
import math
from string import printable
from typing import Any, Dict, List, TYPE_CHECKING

from hypothesis import example, given
from hypothesis.strategies import floats, integers, text
import pytest

from starparse import config, pack, unpack

if TYPE_CHECKING:
    import sys

    if sys.version_info >= (3, 9):
        OrderedDict_ = OrderedDict
    elif sys.version_info >= (3, 7, 2):
        from typing import OrderedDict as OrderedDict_
    else:
        OrderedDict_ = Dict


def parity(packer, unpacker, reference, asserter, packed_reference=None):
    packed = packer(reference)
    if packed_reference is not None:
        asserter(packed, packed_reference)
    unpacked, _ = unpacker(packed)
    asserter(unpacked, reference)


@pytest.mark.parametrize('args, expected', [
    (('4c', b'Alen'), ('Alen', 4)),
    (('4c', b'Alen B'), ('Alen', 4)),
    (('2c', b'Alen B', 2), ('en', 4)),
    (('b', b'\x01\xFF'), (1, 1)),
    (('b', b'\x00\xFF'), (0, 1)),
    (('<i', b'\x00\x88\x00\x00'), (34816, 4))
])
def test_struct(args, expected):
    assert unpack.struct(*args) == expected


@given(n=integers(min_value=0, max_value=2 ** 256))
def test_uint(n):
    packed = pack.uint(n)
    unpacked, offset = unpack.uint(packed)
    assert offset == len(packed)
    assert unpacked == n


@given(n=integers(min_value=-(2 ** 127), max_value=2 ** 127 - 1))
def test_int(n: int):
    packed = pack.int_(n)
    unpacked, offset = unpack.int_(packed)
    assert offset == len(packed)
    assert unpacked == n


@given(string=text(alphabet=printable))
@example(string=printable)
def test_str(string: str):
    packed = pack.str_(string)
    unpacked, offset = unpack.str_(packed)
    assert offset == len(packed)
    assert unpacked == string


@pytest.mark.parametrize('value', [True, False])
def test_bool(value: bool):
    packed = pack.bool_(value)
    unpacked, offset = unpack.bool_(packed)
    assert offset == len(packed)
    assert unpacked == value


def test_none():
    packed = pack.none(None)
    unpacked, offset = unpack.none(packed)
    assert offset == len(packed)
    assert unpacked is None


@pytest.mark.parametrize('n', [-float('nan'), float('nan')])
def test_float_nan(n: float):
    packed = pack.float_(n)
    unpacked, offset = unpack.float_(packed)
    assert offset == len(packed)
    assert math.isnan(unpacked)
    assert math.copysign(1.0, unpacked) == math.copysign(1.0, n)


@pytest.mark.parametrize('n', [-float('nan'), float('nan')])
def test_float_nan(n: float):
    packed = pack.float_(n)
    unpacked, offset = unpack.float_(packed)
    assert offset == len(packed)
    assert math.isnan(unpacked)
    assert math.copysign(1.0, unpacked) == math.copysign(1.0, n)


@given(n=floats(allow_nan=False))
@example(n=float('inf'))
@example(n=-float('inf'))
def test_float(n: float):
    packed = pack.float_(n)
    unpacked, offset = unpack.float_(packed)
    assert offset == len(packed)
    assert unpacked == n


@pytest.mark.parametrize('packed, expected', [
    (b'\x01', type(None)),
    (b'\x02', float),
    (b'\x03', bool),
    (b'\x04', int),
    (b'\x05', str),
    (b'\x06', list),
    (b'\x07', dict),
])
def test_types(packed: bytes, expected: Any, monkeypatch):
    unpacked, offset = unpack.type_(packed)
    assert offset == len(packed)
    assert unpacked == expected


@pytest.mark.parametrize('value', [
    [],
    [1, 2, 3],
    ["a", "b", "c"],
    [1, "b", []],
    [1, "b", [1, "c"]],
])
def test_list(value: List[Any]):
    packed = pack.list_(value)
    unpacked, offset = unpack.list_(packed)
    assert offset == len(packed)
    assert unpacked == value


@pytest.mark.parametrize('value', [
    {},
    {"a": 1, "b": 2},
    {"a": "a", "b": 2},
    {"a": {"b": {}}},
])
def test_dict(value: Dict[Any, Any], monkeypatch):
    monkeypatch.setattr(config, 'ORDERED_DICT', False)

    packed = pack.dict_(value)
    unpacked, offset = unpack.dict_(packed)
    assert offset == len(packed)
    assert unpacked == value


@pytest.mark.parametrize('value', [
    OrderedDict({}),
    OrderedDict({"a": 1, "b": 2}),
    OrderedDict({"a": "a", "b": 2}),
    OrderedDict({"a": OrderedDict({"b": OrderedDict({})})}),
])
def test_dict_ordered(value: 'OrderedDict_[Any, Any]', monkeypatch):
    monkeypatch.setattr(config, 'ORDERED_DICT', True)

    packed = pack.dict_(value)
    unpacked, offset = unpack.dict_(packed)
    assert offset == len(packed)
    assert unpacked == value


starparse/util.py
"""Utility functions."""

import logging
from typing import Any

__all__ = ('diff',)

logger = logging.getLogger(__name__)


def diff(a: Any, b: Any, context: str) -> int:
    """
    Diff two values.

    :param a: First value
    :param b: Second value
    :param context: Context placing the values relative to root values
    :return: Count of diffs
    """
    if isinstance(a, dict) and isinstance(b, dict):
        return _dict_diff(a, b, context=context)
    if isinstance(a, list) and isinstance(b, list):
        return _list_diff(a, b, context=context)
    return _generic_diff(a, b, context=context)


def _dict_diff(a: Any, b: Any, context: str = 'base') -> int:
    a_extra = a.keys() - b.keys()
    b_extra = b.keys() - a.keys()
    diffs = 0
    if a_extra:
        logger.warning(context)
        logger.warning('  extra keys in a: %s', a_extra)
        diffs += len(a_extra)
    if b_extra:
        logger.warning(context)
        logger.warning('  extra keys in b: %s', b_extra)
        diffs += len(b_extra)
    for k in a.keys() & b.keys():
        diffs += diff(a[k], b[k], context + '.' + k)
    return diffs


def _list_diff(a: Any, b: Any, context: str = 'base') -> int:
    if len(a) != len(b):
        logger.warning(context)
        logger.warning('  list len mismatch: %d, %d', len(a), len(b))
        return max(len(a), len(b))
    diffs = 0
    for i, va, vb in zip(map(str, range(len(a))), a, b):
        diffs += diff(va, vb, context + '[' + i + ']')
    return diffs


def _generic_diff(a: Any, b: Any, context: str = 'base') -> int:
    if a != b:
        logger.warning(context)
        logger.warning('  generic mismatch')
        logger.warning('  %s (%s)', a, type(a).__name__)
        logger.warning('  %s (%s)', b, type(b).__name__)
        return 1
    return 0


starparse/unpack.py
"""Unpacking functionality."""

from collections import OrderedDict
from struct import calcsize, unpack_from
from typing import Any, Dict, List, Tuple, Union

from starparse import config

__all__ = ('UnpackingError', 'struct', 'uint', 'int_', 'str_', 'bool_',
           'none', 'float_', 'type_', 'list_', 'dict_', 'typed', 'header')

SBT = Union[None, str, int, float, list, dict, OrderedDict]


class UnpackingError(Exception):
    """Unpacking error."""


def struct(fmt: str, buffer: bytes, offset: int = 0) -> Tuple[Any, int]:
    """
    Unpack struct from Starbound save file.

    :param fmt: struct format
    :param buffer: Starbound save file
    :param offset: Starbound save file format
    :return: data, new offset
    :raises UnpackingError: when format not as expected
    """
    unpacked = unpack_from(fmt, buffer, offset)
    offset += calcsize(fmt)

    # Unpacking string, join all bytes and decode
    if all(isinstance(b, bytes) for b in unpacked):
        try:
            result = b''.join(unpacked).decode('ascii')
        except UnicodeDecodeError as e:
            if config.UTF8:
                result = b''.join(unpacked).decode('utf-8')
            else:
                raise UnpackingError(f'ASCII decoding error {unpacked}') from e
        return result, offset

    # Single type (float, int, ...)
    if len(unpacked) == 1:
        return unpacked[0], offset

    raise UnpackingError('Multiple non-bytes in bytearray')


def uint(buffer: bytes, offset: int = 0) -> Tuple[int, int]:
    """
    Unpack unsigned int from Starbound save file.

    :param buffer: Starbound save file
    :param offset: position in Starbound save file
    :return: unsigned int, new offset
    """
    value = 0
    while True:
        tmp = buffer[offset]
        value = (value << 7) | (tmp & 127)
        offset += 1
        if not tmp & 128:
            break
    return value, offset


def int_(buffer: bytes, offset: int = 0) -> Tuple[int, int]:
    """
    Unpack signed int from Starbound save file.

    :param buffer: Starbound save file
    :param offset: position in Starbound save file
    :return: int, new offset
    """
    value = 0
    while True:
        tmp = buffer[offset]
        value = (value << 7) | (tmp & 127)
        offset += 1
        if not tmp & 128:
            break
    if value & 1:
        value = -((value >> 1) + 1)
    else:
        value >>= 1
    return value, offset


def str_(buffer: bytes, offset: int = 0) -> Tuple[str, int]:
    """
    Unpack str from Starbound save file.

    :param buffer: Starbound save file
    :param offset: position in Starbound save file
    :return: str, new offset
    """
    length, offset = uint(buffer, offset)
    fmt = '{0:d}c'.format(length)
    return struct(fmt, buffer, offset)


def bool_(buffer: bytes, offset: int = 0) -> Tuple[bool, int]:
    """
    Unpack bool from Starbound save file.

    :param buffer: Starbound save file
    :param offset: position in Starbound save file
    :return: bool, new offset
    """
    return bool(buffer[offset]), offset + 1


# pylint: disable=unused-argument
# noinspection PyUnusedLocal
def none(buffer: bytes, offset: int = 0) -> Tuple[None, int]:
    """
    Unpack None/unset from Starbound save file.

    :param buffer: Starbound save file
    :param offset: position in Starbound save file
    :return: None, new offset
    """
    return None, offset


def float_(buffer: bytes, offset: int = 0) -> Tuple[float, int]:
    """
    Unpack float from Starbound save file.

    :param buffer: Starbound save file
    :param offset: position in Starbound save file
    :return: float, new offset
    """
    return struct('>d', buffer, offset)


def type_(buffer: bytes, offset: int = 0) -> Tuple[type, int]:
    """
    Unpack type from Starbound save file.

    :param buffer: Starbound save file
    :param offset: position in Starbound save file
    :return: type, new offset
    :raises UnpackingError: when format not as expected
    """
    types = [type(None), float, bool, int, str, list, dict]
    index, offset = uint(buffer, offset)
    if index > len(types):
        raise UnpackingError(f'Unsupported value type: {index}')
    return types[index - 1], offset


def list_(buffer: bytes, offset: int = 0) -> Tuple[List[SBT], int]:
    """
    Unpack list from Starbound save file.

    :param buffer: Starbound save file
    :param offset: position in Starbound save file
    :return: list, new offset
    """
    length, offset = uint(buffer, offset)
    result = []
    for _ in range(length):
        item, offset = typed(buffer, offset)
        result.append(item)
    return result, offset


def dict_(buffer: bytes, offset: int = 0) -> Tuple[Dict[str, SBT], int]:
    """
    Unpack dict from Starbound save file.

    :param buffer: Starbound save file
    :param offset: position in Starbound save file
    :return: dict, new offset
    """
    length, offset = uint(buffer, offset)

    result: Dict[str, SBT]
    if config.ORDERED_DICT:
        result = OrderedDict()
    else:
        result = {}

    for _ in range(length):
        key, offset = str_(buffer, offset)
        item, offset = typed(buffer, offset)
        result[key] = item

    return result, offset


def typed(buffer: bytes, offset: int = 0) -> Tuple[SBT, int]:
    """
    Unpack a typed data structure from the buffer at a given offset.

    :param buffer: buffer to read
    :param offset: offset in buffer
    :return: unpacked data
    """
    handlers = {
        type(None): none,
        bool: bool_,
        int: int_,
        float: float_,
        list: list_,
        dict: dict_,
        str: str_
    }
    value_type, offset = type_(buffer, offset)
    value, offset = handlers[value_type](buffer, offset)
    return value, offset


def header(buffer: bytes, offset: int = 0) -> Tuple[bytes, str, List[int], int]:
    """
    Unpack a Starbound header structure from the buffer at a given offset.

    :param buffer: buffer to read
    :param offset: offset in buffer
    :return: Starbound header
    """
    save_format = buffer[offset:offset + 6]
    offset += 6
    entity, offset = str_(buffer, offset=offset)
    flags = list(buffer[offset:offset + 5])
    offset += 5
    return save_format, entity, flags, offset


starparse/pack.py
"""Packing functionality."""

from collections import OrderedDict
from functools import wraps
from struct import pack
from typing import Any, Callable, Dict, List, TypeVar, Union, get_type_hints

from starparse import config

T = TypeVar('T')
SBT = Union[str, int, float, list, dict, OrderedDict]

__all__ = ('PackingError', 'uint', 'int_', 'str_', 'bool_', 'none',
           'float_', 'type_', 'list_', 'dict_', 'typed', 'header')


class PackingError(Exception):
    """Packing error."""


def check_type(f: Callable[[T], bytearray]) -> Callable[[T], bytearray]:
    """
    Check function argument type.

    :param f: function to check param value for
    :return: function with param value checking
    """
    @wraps(f)
    def wrapper(value):
        expecting = get_type_hints(f)['value']
        if (expecting.__module__ == 'typing'
                and expecting.__origin__ in (list, List)):
            expecting = list
        elif (expecting.__module__  == 'typing'
              and expecting.__origin__ in (dict, Dict)):
            if config.ORDERED_DICT:
                expecting = OrderedDict
            else:
                expecting = dict
        if not isinstance(value, expecting):
            raise TypeError(f'{f.__module__}.{f.__name__} expecting '
                            f'{expecting.__name__} but got '
                            f'{type(value).__name__} ({value!r})')

        return f(value)

    return wrapper


@check_type
def uint(value: int) -> bytearray:
    """
    Pack type to Starbound format.

    :param value: unsigned int
    :return: bytearray
    :raises PackingError: when int negative
    """
    if value < 0:
        raise PackingError(f'unsigned int cannot be negative: {value}')
    result = bytearray()
    result.insert(0, value & 127)
    value >>= 7
    while value:
        result.insert(0, value & 127 | 128)
        value >>= 7
    return result


@check_type
def int_(value: int) -> bytearray:
    """
    Pack int to Starbound format.

    :param value: int
    :return: bytearray
    """
    value_ = abs(value * 2)
    if value < 0:
        value_ -= 1
    return uint(value_)


@check_type
def str_(value: str) -> bytearray:
    """
    Pack string to Starbound format.

    :param value: string
    :return: bytearray
    :raises PackingError: when string encoding error
    """
    result = uint(len(value))
    try:
        result.extend(bytearray(value, 'ascii'))
    except UnicodeEncodeError as e:
        if config.UTF8:
            result.extend(bytearray(value, 'utf-8'))
        else:
            raise PackingError(f'string encoding error: {value!r}') from e
    return result


@check_type
def bool_(value: bool) -> bytearray:
    """
    Pack bool to Starbound format.

    :param value: bool
    :return: bytearray
    """
    return bytearray([value])


# pylint: disable=unused-argument
# noinspection PyUnusedLocal
def none(value: Any = None) -> bytearray:
    """
    Pack None/unset to Starbound format.

    :param value: unused
    :return: bytearray
    """
    return bytearray()


@check_type
def float_(value: float) -> bytearray:
    """
    Pack float to Starbound format.

    :param value: float
    :return: bytearray
    """
    return bytearray(pack('>d', value))


def type_(value: type) -> bytearray:
    """
    Pack type to Starbound format.

    :param value: type
    :return: bytearray
    :raises PackingError: when unsupported value type
    """
    types = dict(zip((type(None), float, bool, int, str, list, dict),
                     range(1, 8)))
    types[OrderedDict] = types[dict]
    try:
        value_type = types[value]
    except KeyError as e:
        raise PackingError(f'unsupported value type: {value}') from e
    return uint(value_type)


@check_type
def list_(value: List[SBT]) -> bytearray:
    """
    Pack list to Starbound format.

    :param value: type
    :return: bytearray
    """
    result = uint(len(value))
    for val in value:
        result.extend(typed(val))
    return result


@check_type
def dict_(value: Dict[str, SBT]) -> bytearray:
    """
    Pack dict to Starbound format.

    :param value: type
    :return: bytearray
    """
    result = uint(len(value))
    for key, val in value.items():
        result.extend(str_(key))
        result.extend(typed(val))
    return result


def typed(value: SBT) -> bytearray:
    """
    Pack type and value to Starbound format.

    :param value: value
    :return: bytearray
    """
    handlers: Dict[type, Callable[[Any], bytearray]] = {
        type(None): none,
        bool: bool_,
        int: int_,
        float: float_,
        list: list_,
        dict: dict_,
        OrderedDict: dict_,
        str: str_
    }
    result = type_(type(value))
    result.extend(handlers[type(value)](value))
    return result


def header(save_format: bytes, entity: str, flags: List[int]) -> bytearray:
    """
    Pack Starbound header to Starbound format.

    :param save_format: save format
    :param entity: entity
    :param flags: flags
    :return: bytearray
    """
    return bytearray(save_format) + str_(entity) + bytearray(flags)
